### PR TITLE
[HTML] Fallback and Requires support

### PR DESCRIPTION
**Pull request by dclaux**
Implements https://github.com/Microsoft/AdaptiveCards/issues/2019

### CODE REVIEW

**Comment by dclaux** [comment]
Sample payload to test with:

```
{
    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
    "type": "AdaptiveCard",
    "version": "1.0",
    "fallbackText": "This is the fallback text",
    "fallback": {
        "type": "TextBlock",
        "text": "This is the fallback element"
    },
    "body": [
        {
            "type": "Container",
            "fallback": {
                "type": "TextBlock",
                "text": "The container is falling back."
            },
            "selectAction": {
                "type": "Action.OpenUrl",
                "title": "Hello world",
                "url": "http://www.microsoft.com"
            },
            "items": [
                {
                    "type": "TextBlock",
                    "text": "Publish Adaptive Card schema",
                    "weight": "bolder",
                    "size": "medium"
                },
                {
                    "type": "ActionSet",
                    "actions": [
                        {
                            "type": "Action.OpenfUrl",
                            "title": "Hello world",
                            "url": "http://www.microsoft.com"
                        }
                    ]
                },
                {
                    "type": "Image",
                    "size": "large",
                    "url": "http://cdn.nirmaltv.com/images/Powerpoint-2013.png",
                    "fallback": "drop",
                    "selectAction": {
                        "type": "Action.OpenUrl",
                        "title": "Hello world",
                        "url": "http://www.microsoft.com"
                    }
                }
            ]
        },
        {
            "type": "ColumnSet",
            "selectAction": {
                "type": "Action.OpenUrl",
                "title": "Hello world",
                "url": "http://www.microsoft.com"
            },
            "columns": [
                {
                    "width": 1,
                    "style": "emphasis",
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "Column 1",
                            "horizontalAlignment": "center"
                        }
                    ]
                },
                {
                    "width": 1,
                    "style": "emphasis",
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "Column 2",
                            "horizontalAlignment": "center"
                        }
                    ]
                },
                {
                    "width": 1,
                    "style": "emphasis",
                    "fallback": "drop",
                    "items": [
                        {
                            "type": "TextBlock",
                            "text": "Column 3",
                            "horizontalAlignment": "center"
                        }
                    ]
                }
            ]
        }
    ],
    "actions": [
        {
            "type": "Action.OpefnUrl",
            "title": "Hello world",
            "url": "http://www.microsoft.com",
            "fallback": {
                "type": "Action.OpenUrl",
                "title": "Fallback action",
                "url": "http://www.microsoft.com"
            }
        }
    ]
}
```

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 1]
```diff
@@ -26,19 +26,101 @@ function generateUniqueId(): string {
...
+            if (typeof fallback === "string" && fallback.toLowerCase() === "drop") {
+                result = null;
+            }
+            else if (typeof fallback === "object") {
+                result = createElementInstance(
+                    parent,
+                    fallback,
+                    errors);
+            }
+        }
```
Comment (paulcam206) [inline source/nodejs/adaptivecards/src/card-elements.ts]: a lot of this logic is duplicated from above. helper function?

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 2]
```diff
@@ -3325,6 +3412,23 @@ class ActionCollection {
         this._owner = owner;
     }
 
+    parse(json: any, errors?: Array<IValidationError>) {
```
Comment (paulcam206) [inline source/nodejs/adaptivecards/src/card-elements.ts]: >         [](start = 56, length = 8)

looks like you haven't configured your editor to consume `.editorconfig` yet? instructions [here](https://editorconfig.org/#download).

**Comment by paulcam206** [comment]
>             throw new Error("The action already belongs to another element.")

or is null? also, `;`

---
Refers to: source/nodejs/adaptivecards/src/card-elements.ts:3666 in 077c274. [](commit_id = 077c274d9d8be96db7f6948b3221a21e0d1d1815, deletion_comment = False)

File: source/nodejs/adaptivecards/src/card-elements.ts (hunk 2 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts]: Done.

File: source/nodejs/adaptivecards/src/card-elements.ts (hunk 1 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts]: Done

**Comment by dclaux** [comment]
Now supports "requires" clause. Can be tested via the Outlook host in the visualizer, which defines two "capabilities":
- helloWorld, any version
- whatNot, version 2.12

To test, select the Outloook Actionable Messages host container, and add a "requires" property to any element, action or the card itself. For instance:
```
"requires": {
  "whatNow": "3.0"
}
```

The above will cause the element/action/card to fall back.

File: source/nodejs/adaptivecards/src/host-config.ts [hunk 3]
```diff
@@ -330,7 +330,166 @@ export class ContainerStyleSet {
     }
 }
 
+export class Version {
+    private _versionString: string;
+    private _major: number;
+    private _minor: number;
```
Comment (paulcam206) [inline source/nodejs/adaptivecards/src/host-config.ts]: shouldn't we be supporting semantic versioning? there should be a `build` and `revision` property

File: source/nodejs/adaptivecards/src/host-config.ts [hunk 4]
```diff
@@ -330,7 +330,166 @@ export class ContainerStyleSet {
...
+                break;
+            }
+        }
+
+        minorVersionString = minorVersionString.substr(0, minorVersionString.length - trailingZeros);
+
+        return !this._isValid ? this._versionString : this._major + "." + minorVersionString;
+    }
+
+    compareTo(otherVersion: Version): number {
```
Comment (paulcam206) [inline source/nodejs/adaptivecards/src/host-config.ts]: > compareTo(otherVersion: Version): [](start = 3, length = 34)

looks like this won't work for wildcards or version ranges (per spec)

File: source/nodejs/adaptivecards/src/host-config.ts (hunk 3 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/host-config.ts]: There is an active discussion on the side about this. I think we should keep it simple, but I will update the code according to the conclusion of the conversation.

File: source/nodejs/adaptivecards/src/host-config.ts (hunk 4 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/host-config.ts]: My implementation doesn't support ranges at the moment, and uses the explicit string "any" for wildcard. There is an active discussion on the side about versioning. I think we should keep things simple, but I will update the code according to the conclusion of the conversation.

Comment (dclaux) [reply source/nodejs/adaptivecards/src/host-config.ts]: Per discussion, we are going with a simple model for 1.2.

File: source/nodejs/adaptivecards/src/host-config.ts (hunk 3 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/host-config.ts]: Per discussion we are going with a simple model for 1.2.

**Pull request merged**